### PR TITLE
Fix Delta vacuum when location with trailing slash

### PR DESCRIPTION
**Pull request by pajaks**
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by ebyhr** [comment]
/test-with-secrets sha=f211b890d417022baccf80ab2bd78c1e668dc255

File: lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java [hunk 1]
```diff
@@ -68,11 +68,14 @@ public FileEntry next()
         }
```
Comment (findepi) [inline lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java:68]: > Fix listening files with trailing slash location

listening -> listing

File: lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java [hunk 2]
```diff
@@ -68,11 +68,14 @@ public FileEntry next()
...
 
         String root = listingPath.toUri().getPath();
+        if (!root.endsWith("/")) {
+            root += "/";
```
Comment (findepi) [inline lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java:72]: `listingPath` is const, so is `root`.
Can this logic be applied once in the ctor?

Also, add a comment why append slash, what cases this is needed for.

File: lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java [hunk 3]
```diff
@@ -68,11 +68,14 @@ public FileEntry next()
...
+        }
         String path = status.getPath().toUri().getPath();
```
Comment (findepi) [inline lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java:74]: `status.getPath().toUri()` -- does it do normalization when path contains repeated slashes?
do we have test covering file listings from locations with double slashes?

File: lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java [hunk 4]
```diff
@@ -68,11 +68,14 @@ public FileEntry next()
...
 
         verify(path.startsWith(root), "iterator path [%s] not a child of listing path [%s] for location [%s]", path, root, listingLocation);
 
-        Location location = listingLocation.appendPath(path.substring(root.length() + 1));
+        Location location = listingLocation.appendPath(path.substring(root.length()));
```
Comment (findepi) [inline lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java:78]: I don't understand this logic: we're stripping `root` prefix (which comes from `listingPath`) to prepend `listingPath`. does it yield `path` back? or something else?
let's have explanatory comment

**Comment by findepi** [comment]
@losipiuk does this fix the problem with top-of-bucket listing you found last week?

File: lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java (hunk 2 above)
Comment (losipiuk) [reply lib/trino-hdfs/src/main/java/io/trino/filesystem/hdfs/HdfsFileIterator.java:72]: Not sure but it does not look so.

The issue I observed was coming from the fact that here:
https://github.com/trinodb/trino/blob/5b8e06a9ecfaf5b2e7255419b787c00f9c0c345d/plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/DirectoryListingFilter.java#L71

we got parent directory of a file without trailing slash. While `prefix` passed to `DirectoryListingFilter` was with slash - and filtering did not work as expected.

Will this code path be affected?

**Comment by losipiuk** [comment]
> @losipiuk does this fix the problem with top-of-bucket listing you found last week?

I verified locally with Trino built with this PR and it dos not help.

**Comment by findinpath** [comment]
@pajaks  can you please rebase on master to make use of #17848 ?

**Pull request closed without merging**